{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# W200 Introduction to Data Science Programming, UC Berkeley MIDS\n",
    "\n",
    "\n",
    "## Instructions\n",
    "The final exam is designed to evaluate your grasp of Python theory as well as Python coding.\n",
    "\n",
    "- This is an individual exam.\n",
    "- You have 48 hours to complete the exam, starting from the point at which you first access it.\n",
    "- You will be graded on the quality of your answers.  Use clear, persuasive arguments based on concepts we covered in class.\n",
    "- Please double-click the markdown cells where it says \"Your answer here\" to input answers (if you need more cells please make them markdown cells)\n",
    "- Use only Python standard libraries, matplotlib, seaborn, NumPy and Pandas for this exam\n",
    "- Please push the exam to your github repo in the folder /SUBMISSIONS/final_exam/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Ahmad Azizi"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1: Short Answer Questions (25 pts - each question = 5 pts)\n",
    "\n",
    "a) The following class ```Cart``` and method ```add_to_cart``` are parts of a larger program used by a mobile phone company.  The method ```add_to_cart``` will work when an object of type MobileDevice or of type ServiceContract is passed to it. State whether the method ```add_to_cart``` is a demonstration of the following items (yes/no) and the reasoning (1-2 sentences): \n",
    "\n",
    "    1. Inheritance\n",
    "    2. Polymorphism\n",
    "    3. Duck typing\n",
    "    4. Top-down design\n",
    "    5. Functional programming"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Method:\n",
    "class Cart():\n",
    "    \n",
    "    def __init__(self):\n",
    "        self.cart = []\n",
    "        self.total = 0\n",
    "    \n",
    "    def add_to_cart(self, item):\n",
    "        self.cart.append(item)\n",
    "        self.total += item.price"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1- Inheritance: No\\\n",
    "Reasoning: There is a class named \"cart\" and we don't see any class inheriting this class. \n",
    "\n",
    "2- Polymorphism: No\\\n",
    "Reasoning: Because in polymorphism, we need to have another function with the same function name but with different parameters. In the above example, we only see one add_to_cart function.\n",
    "\n",
    "3- Duck Typing: Yes (read note for No)\\\n",
    "Reasoning: In Duck Typing, the type or class of an object is less important that the method it defines. In other words, you do not check for type at all, you check for a given method.\\\n",
    "\n",
    "Note: if we read the description above, it does mention that \"object of type MobileDevice or of type ServiceContract is passed to it.\" In that case, it would be a No. (I wasn't sure whehter to explicitly choose yes or no)\n",
    "\n",
    "4- Top-down Design: Yes\\\n",
    "Reasoning: In top-down design, the problem is broken down into smaller pieces. The description mentions that this class part of a larger program. Therefore, it is part of a top-down design. \n",
    "\n",
    "5- Functional Programming: Yes\\\n",
    "Reasoning: because this program uses object oriented approach. \n",
    "\n",
    "1. Inheritance:\n",
    "2. Polymorphism:\n",
    "3. Duck typing:\n",
    "4. Top-down design:\n",
    "5. Functional programming:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "b) Suppose you have a long list of digits (0-9) that you want to write to a file.  From a storage standpoint, would it be more efficient to use ASCII or UTF-8 as an encoding? What is the most efficient way to create an even smaller file to store the information?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I would use ASCII in this situation since it is only representing numbers. From a storage standpoint, UTF-8 starts with 8 bits and ASCII with 7 bits. Choosing ASCII will ensure that we save 1 bit for each character.\\ \n",
    "\n",
    "We could perhaps use hexadecimal to store our digits instead of storing digit as a separater character. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "c) Why is it important to sanity-check your data before you begin your analysis? What could happen if you don't?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "Sanity checks are performed to ensure our code is performing as we expect it to. For example, in the above code, we perform a sanity check to ensure the data does not cause any errors, to account for any mismatch values, missing valuse, etc. If we do not perform a sanity check, imagine running a large dataset only to find that some of the phonenumbers have dashes in between them, in that case, we would hit a wall and have to re-adjust our code.  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "d) How do you determine which variables in your dataset you should check for issues prior to starting an analysis? "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "This question is very broad and will vary based on different situations. I would divide my datasets variables into two catergories, main variables and supplemental variables.\\\n",
    "\n",
    "For example, in a dataset that breaks down the GDP of countries based on specific industries, I would make the country name as a main variable and the rest of industry specific information as supplemental variables. If the name of the country is missing, I would get rid of the rest of supplemental information, because if we do not know the name of the country then the rest of the data is useless. If the name of the country exists in our dataset and some of the supplemental data is missing, I would include the that country in my dataset. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "e1) Explain why the following code prints what it does."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'function'>\n"
     ]
    }
   ],
   "source": [
    "def f(): pass\n",
    "print(type(f))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "-Defines a function f. The function f() does nothing and returns nothing.\\\n",
    "-uses Python built in function type() and plugs in f as an argument. Prints the result, which is a fucntion type. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "e2) Explain why the following code prints something different."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'NoneType'>\n"
     ]
    }
   ],
   "source": [
    "def f(): pass\n",
    "print(type(f()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "-Defines a function f. The function f() does nothing and returns nothing.\\\n",
    "-Uses Python built in function type(), but this time it plugs in f() as an argument and since f() does not return anything, the code will print non object because it has no value and hence NonType. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2:  General Coding Questions (15 pts - each question 5 pts)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "a) Using a list comprehension: Make a list of the squared numbers greater than 25 that are the square of non-negative integer less than 10. Fill in a list comprehension below so that we get this desired output."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[36, 49, 64, 81]\n"
     ]
    }
   ],
   "source": [
    "my_list = [i*i for i in range(10) if i*i > 25 ]\n",
    "print(my_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "b) Below is a data frame of customers that have different cooling systems. Your data science team lead wants the column ```cooling_system``` to be labeled with the integer numbers 1-4 instead of the text as shown below:\n",
    "\n",
    "- 1 = Air Conditioning / AC / Air Con\n",
    "- 2 = Heat Pump / HP\n",
    "- 3 = Evaporative Cooler / Evap Cooler\n",
    "- 4 = Fan\n",
    "\n",
    "Make a new column called ```cooling_type``` that maps the text values to the new numeric values. Filter out the values that are not included in the mapping above. Print out/display this new data frame. Be sure to list any assumptions also!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cust_Number</th>\n",
       "      <th>Cooling_System</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>101</td>\n",
       "      <td>AC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>102</td>\n",
       "      <td>Heat Pump</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>103</td>\n",
       "      <td>Air Con</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>104</td>\n",
       "      <td>Air Conditioning</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>105</td>\n",
       "      <td>Fan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>106</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>107</td>\n",
       "      <td>Evap Cooler</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>108</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>109</td>\n",
       "      <td>AC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>110</td>\n",
       "      <td>Evaporative Cooler</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>111</td>\n",
       "      <td>geothermal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>112</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Cust_Number      Cooling_System\n",
       "0           101                  AC\n",
       "1           102           Heat Pump\n",
       "2           103             Air Con\n",
       "3           104    Air Conditioning\n",
       "4           105                 Fan\n",
       "5           106                None\n",
       "6           107         Evap Cooler\n",
       "7           108                None\n",
       "8           109                  AC\n",
       "9           110  Evaporative Cooler\n",
       "10          111          geothermal\n",
       "11          112                   1"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas\n",
    "\n",
    "# creating a data frame from scratch - list of lists\n",
    "\n",
    "data = [ [101, 'AC'], \n",
    "         [102, 'Heat Pump'], \n",
    "         [103, 'Air Con'], \n",
    "         [104, 'Air Conditioning'],\n",
    "         [105, 'Fan'],\n",
    "         [106, 'None'],\n",
    "         [107, 'Evap Cooler'],\n",
    "         [108, None],\n",
    "         [109, 'AC'],\n",
    "         [110, 'Evaporative Cooler'],\n",
    "         [111, 'geothermal'],\n",
    "         [112, 1]\n",
    "       ]\n",
    "\n",
    "# create a data frame with column names - list of lists\n",
    "\n",
    "col_names = ['Cust_Number', 'Cooling_System']\n",
    "df = pandas.DataFrame(data, columns=col_names)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cust_Number</th>\n",
       "      <th>Cooling_System</th>\n",
       "      <th>cooling_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>101</td>\n",
       "      <td>AC</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>102</td>\n",
       "      <td>Heat Pump</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>103</td>\n",
       "      <td>Air Con</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>104</td>\n",
       "      <td>Air Conditioning</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>105</td>\n",
       "      <td>Fan</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>107</td>\n",
       "      <td>Evap Cooler</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>109</td>\n",
       "      <td>AC</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>110</td>\n",
       "      <td>Evaporative Cooler</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Cust_Number      Cooling_System  cooling_type\n",
       "0          101                  AC             1\n",
       "1          102           Heat Pump             2\n",
       "2          103             Air Con             1\n",
       "3          104    Air Conditioning             1\n",
       "4          105                 Fan             4\n",
       "6          107         Evap Cooler             3\n",
       "8          109                  AC             1\n",
       "9          110  Evaporative Cooler             3"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def finding_cooling_type (i):\n",
    "    \n",
    "    if i == 'Air Conditioning' or i == 'AC' or i == 'Air Con':\n",
    "        return 1\n",
    "    \n",
    "    if i == 'Heat Pump' or i == 'HP':\n",
    "        return 2\n",
    "    \n",
    "    if i == 'Evaporative Cooler' or i == 'Evap Cooler':\n",
    "        return 3\n",
    "    \n",
    "    if i == 'Fan':\n",
    "        return 4\n",
    "    \n",
    "df['cooling_type'] = df['Cooling_System'].apply(finding_cooling_type)\n",
    "df.dropna(inplace = True)\n",
    "df['cooling_type'] = df['cooling_type'].astype(int)\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "c) From the dataframe below, use groupby in Pandas to show how many total delegates were obtained grouped by favorite color. Print this out."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>delegates</th>\n",
       "      <th>color</th>\n",
       "      <th>state</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>marco</td>\n",
       "      <td>165</td>\n",
       "      <td>blue</td>\n",
       "      <td>FL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>jeb</td>\n",
       "      <td>0</td>\n",
       "      <td>red</td>\n",
       "      <td>FL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chris</td>\n",
       "      <td>0</td>\n",
       "      <td>white</td>\n",
       "      <td>NJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>donald</td>\n",
       "      <td>1543</td>\n",
       "      <td>white</td>\n",
       "      <td>NY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ted</td>\n",
       "      <td>559</td>\n",
       "      <td>blue</td>\n",
       "      <td>TX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>john</td>\n",
       "      <td>161</td>\n",
       "      <td>red</td>\n",
       "      <td>OH</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     name  delegates  color state\n",
       "0   marco        165   blue    FL\n",
       "1     jeb          0    red    FL\n",
       "2   chris          0  white    NJ\n",
       "3  donald       1543  white    NY\n",
       "4     ted        559   blue    TX\n",
       "5    john        161    red    OH"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas\n",
    "\n",
    "# creating a data frame from scratch - list of lists\n",
    "\n",
    "data = [ ['marco', 165, 'blue', 'FL'], \n",
    "         ['jeb', 0, 'red', 'FL'], \n",
    "         ['chris', 0, 'white', 'NJ'], \n",
    "         ['donald', 1543, 'white', 'NY'],\n",
    "         ['ted', 559, 'blue', 'TX'],\n",
    "         ['john', 161, 'red', 'OH']\n",
    "       ]\n",
    "\n",
    "# create a data frame with column names - list of lists\n",
    "\n",
    "col_names = ['name', 'delegates', 'color', 'state']\n",
    "df = pandas.DataFrame(data, columns=col_names)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>delegates</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>color</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>blue</th>\n",
       "      <td>724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>red</th>\n",
       "      <td>161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>white</th>\n",
       "      <td>1543</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       delegates\n",
       "color           \n",
       "blue         724\n",
       "red          161\n",
       "white       1543"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.groupby(['color']).sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3: Bitcoin coding problem (20 points):"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Bitcoin\n",
    "\n",
    "Consider a record of a one-time investment in bitcoin with value of that investment tracked monthly, provided as an (ordered) tuple of dictionaries, where each dictionary comprises one key for the month and corresponding one value for the value of the investment, and the first entry (Jan 2018) is the initial investment made on 01 Jan 2018, shown in `data` below.\n",
    "\n",
    "Write Python code to take such a record of any length (the below <i>data</i> is only a sample), and output a table/dataframe comprising a row for each month with columns for date, start balance, and return. **Print out this table/dataframe.**\n",
    "\n",
    "Also, visualize the record as two vertically arranged plots. \n",
    "- The top plot should show a line plot of start balance vs. month\n",
    "- The bottom plot should show a bar plot of return vs. month, with a black horizontal line at return=0, and bars color-coded such that positive returns are green and negative returns are red.  \n",
    "- The two plots' horizontal axes should align. Demonstrate that your code works by applying it to `data`.\n",
    "\n",
    "Notes: <br>\n",
    "- The gain for each period is the end balance minus the start balance. \n",
    "- The growth factor for each period is the end balance divided by the start balance.  \n",
    "- The return for each period is the growth factor minus 1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "({'Jan 2018': 1000},\n",
       " {'Feb 2018': 1100},\n",
       " {'Mar 2018': 1400},\n",
       " {'Apr 2018': 700},\n",
       " {'May 2018': 800},\n",
       " {'Jun 2018': 500})"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = ({\"Jan 2018\":1000},{\"Feb 2018\":1100},{\"Mar 2018\":1400},{\"Apr 2018\":700},{\"May 2018\":800},{\"Jun 2018\":500})\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>date</th>\n",
       "      <th>start_balance</th>\n",
       "      <th>return</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Jan 2018</td>\n",
       "      <td>1000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Feb 2018</td>\n",
       "      <td>1100</td>\n",
       "      <td>0.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Mar 2018</td>\n",
       "      <td>1400</td>\n",
       "      <td>0.27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Apr 2018</td>\n",
       "      <td>700</td>\n",
       "      <td>-0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>May 2018</td>\n",
       "      <td>800</td>\n",
       "      <td>0.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Jun 2018</td>\n",
       "      <td>500</td>\n",
       "      <td>-0.38</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       date start_balance return\n",
       "0  Jan 2018          1000    0.0\n",
       "1  Feb 2018          1100    0.1\n",
       "2  Mar 2018          1400   0.27\n",
       "3  Apr 2018           700   -0.5\n",
       "4  May 2018           800   0.14\n",
       "5  Jun 2018           500  -0.38"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "\n",
    "data = ({\"Jan 2018\":1000},{\"Feb 2018\":1100},{\"Mar 2018\":1400},{\"Apr 2018\":700},\n",
    "        {\"May 2018\":800},{\"Jun 2018\":500})\n",
    "\n",
    "\n",
    "month = []\n",
    "balance = []\n",
    "Return = []\n",
    "last_balance = -1\n",
    "\n",
    "\n",
    "for i in data:\n",
    "    for key, val in i.items():\n",
    "        month.append(key)\n",
    "        balance.append(val)\n",
    "        \n",
    "        if last_balance != -1:\n",
    "            R = (val/last_balance) - 1\n",
    "            R = round(R, 2)\n",
    "        else:\n",
    "            R = 0.0\n",
    "            \n",
    "        last_balance = val\n",
    "        Return.append(R)\n",
    "        \n",
    "my_data = pd.DataFrame([month, balance, Return]).T\n",
    "my_data.columns = ['date', 'start_balance', 'return']\n",
    "my_data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZAAAAFDCAYAAAAUFVhNAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAABCkElEQVR4nO3dd5xU1fnH8c8XkM7SRaorglT7il2wNxRssUWNJcb8VGwpamI0lsTE3hNjN9ZYolEUFXtBBcECSBFBcFGkSJf6/P44Z3BYF5idnd07s/u8X6997c6dO7PP3Zmd557znHuOzAznnHOuouokHYBzzrnC5AnEOedcVjyBOOecy4onEOecc1nxBOKccy4rnkCcc85lxROIyxlJ/5B0SRU87wuSTsr18xYaSV0kLZJUtwp/xwBJM9Juj5U0oAp+zyJJXXP9vK56eQJxGZM0VdLS+M8/T9Lzkjqn7jezM8zsirjvWh9ElWFmB5rZ/TmI+RtJ90lqmuFjX5d0Wja/t6IkNZT0vaS9yrnvBklPmNlXZtbUzFZVR0wAZtbHzF6vzHOU93eMxzGlUsG5xHkCcRV1iJk1BdoD3wK3JBxPJlIxbwNsC1xUHb9UUr1M9zWzH4DHgBPLPEdd4FggqwTqXFXyBOKyEj/wngB6p7bFs/srJTUBXgA6xDP/RZI6SKor6WJJX0haKGlUqgUjaRdJH0qaH7/vkva8a85gJf1C0tuSro2toC8lHZhhzN8AwwiJJPXcO0l6N579f5zqrpF0FbA7cGuM/1ZJxZIsPTGUE9s7scUwF7gs/k1ui621hZLel7T5OkK8HzhCUuO0bfsT/k9fKPv74++bEp/3S0nHx+2XSfp3WoxlH3eypPHxcVMk/Wpdf7PYgtsn/vx92uu5OD5nsaSWkp6T9F18TZ6T1Gldf8e43SR1iz83l/RAfPw0SX+UVCftGLN6vV3V8wTishI/5I4GRpS9z8wWAwcCpbGroqmZlQLnE86mDwKKgFOAJZJaAc8DNwOtgeuB5yW1Xsev3xGYALQB/g7cLUkZxNwpxjU53u4Yf++VQCvgN8CTktqa2R+At4CzYvxnZfBnScU2BdgYuCpuOxb4M9Ay/u6rynugmb0LzAQOT9t8AvCwma0scyxNCH+vA82sGbALMCbDGGcBAwmvwcnADZK229CDzKxF6vUEbiL8fb4mfI7cC2wKdAGWArfGx2Tyd7wFaA50BfoTWmEnp92f1evtqp4nEFdR/5X0PbAA2Be4pgKPPQ34o5lNsOBjM5sDHAxMMrMHzWylmT0CfA4cso7nmWZm/4q1gPsJ3WntNhDzQmA64cPz0rj958BQMxtqZqvN7GVgJCHBZavUzG6Jx7E0bnvKzD6ISeAh0lpA5XiA2I0lqQgYxLq7r1YDfSU1MrOZZjY2kwDN7Hkz+yK+Bm8ALxFaCRmRdDRwHHCEma0wszlm9qSZLTGzhYQE2T/D56pLOBG5yMwWmtlU4DpC4kyp6OvtqoknEFdRg82sBdAAOAt4Q9ImGT62M/BFOds7ANPKbJsGdFzH83yT+sHMlsQf11cYHxzP0gcAPQlnshDOmI+KXTPfx8S4G+EDKlvT1xcvsGQDsT4A7BlbR0cCk81sdNmdYivvaOAMYGbsIuuZSYCSDpQ0QtLceMwH8ePfZEOP3ZbQujjMzL6L2xpL+mfsfloAvAm0UGajxdoA9Vn79S/72lf09XbVxBOIy4qZrTKzp4BVhA/dn+xSzrbpQHn9/6WED/N0XQjdIzkTz7bvA65Ni+fB2DWT+mpiZlenHlLmKRbH7+k1irLJs1LTW5vZV4Qun+MJZ+EPrGffYWa2LyHhfQ78Ky3OcmOU1AB4kvA3aBdPBoYCmXQBtgWeJnRHpSe1C4AewI5mVgTskXpIKtT1PO1sYAVrv/45f+1d1fAE4rKiYBChX398Obt8C7SW1Dxt213AFZK6x8dvFescQ4EtJB0nqV7sIukNPFcFod8I7CtpG+DfwCGS9lco8DdUGH7cKe0Y1lyrEM+4vwZ+Hvc/hfITYmXdT2jd7Uro8voJSe0kHRprIcuARYRkDqEWsofCdSPNWXvUWX1C6/E7YGUsSO+3oYBiAf5J4CEze6zM3c0IdY/vYz3r0jL3r/V3TBe7pR4HrpLUTNKmhFrZv8vb3+UXTyCuov4naRGhBnIVcFJ5fe9m9jnwCDAldg91IBTHHyf0uS8A7gYaxTrIQMKZ7Bzgd8BAM5ud6+BjEngAuMTMphNqDBcTPlCnA7/lx/+Lm4Aj4+ifm+O2X8Z95gB9gHdzHSNhdFtLYLiZzVzHPnUIf69SYC6h5vB/ALGW8xjwCTCKtEQcaxRDCK/DPEIt49kMYupEqJOcmzYSa5GkLoSk3IjQmhgBvFjmseX9HdOdTWg1TQHeBh4G7skgJpcw+YJSzjnnsuEtEOecc1nxBOKccy4rnkCcc85lxROIc865rHgCcc45lxVPIM4557LiCcQ551xWPIE455zLiicQ55xzWfEE4pxzLiueQJxzzmXFE4hzzrmseAJxzjmXFU8gzjnnsuIJxDnnXFY8gTjnnMuKJxDnnHNZqZd0ANWpTZs2VlxcnHQYzjlXUEaNGjXbzNqW3V6rEkhxcTEjR45MOgznnCsokqaVt927sJxzzmXFE4hzzrms1KouLOcy9f2S5dz22mSWr1zNpYf0oU4dJR2Sc3nHE4hzaVasWs3D73/FDa9M5PslKwDo2LIRp++xecKROZd/vAvLueiNid9x4E1vcemzY+m1SRFDh+zOQVtuwt9enMCoaXOTDs+5vOMJxNV6k2ct4uR7P+Ckez5gxarV3HnC9jz8yx3p3aGIq4/Yio4tGnHWw6OZt3h50qE6l1c8gbha6/sly7ns2bHsf+ObjJw6j4sP6slL5+3Bfn02QQo1j6KGG3HbcdsxZ9FyLvjPx6xebQlH7Vz+8BqIq3VWrFrNQyOmccMrk1j4wwqO6deF8/fdgjZNG5S7/5admvPHgb340zNj+ddbU/hVf6+HOAeeQFwt89qEWVz53Di++G4xu2zemksG9qZX+6INPu6EnTbl/Slz+fuwCWy/aUtKiltVQ7TO5TfvwnK1wuRZC/nFvR9w8r0fsmq18a8TS3jotB0zSh4AkvjrEVvSqWUjzn5kNHO9HuKcJxBXs81bnKpzvMWoqfP4w0G9eOm8/uzbu92aOkem1qqHPD7G6yGu1vMuLFcjrVi1mn+PmMaNsc5xbKxztF5HnSNTfTs255KBvbjkmbHc+dYUzvB6iKvFPIG4Gue1z2dxxfPjmPLdYnbr1oY/DuxFz00y66rKxM932pQRU+ZyzbAJlHg9xNVi3oXlaoxJ3y7kxHs+4OT7PsQM7jqxhAdP7ZfT5AGhHnJ1rIec9bDXQ1ztVe0JRNI9kmZJ+ixt2zWSPpf0iaSnJbWI24slLZU0Jn79I+0x20v6VNJkSTeroh3arsaYt3g5lz7zGQfc9Bajv5rHHw/uxbBz92CfLOocmWoW6yFzFy/nfK+HuFoqiRbIfcABZba9DPQ1s62AicBFafd9YWbbxK8z0rbfAZwOdI9fZZ/T1XArVq3mnre/pP81r/HgiGkc168Lb/x2T07bvSv161X9W7tvx+ZcckhvXp/wHf98c0qV/z7n8k1WNRBJ7YC/AB3M7EBJvYGdzezuDT3WzN6UVFxm20tpN0cAR27g97cHiszsvXj7AWAw8EJFjsMVJjML13M8P54p3y1m9+5t+OPBvemxSbNqj+XnO3ZhxJQ5XPvSBEqKW7KD10NcLZLtadp9wDCgQ7w9ETg3B/EAnMLaiWAzSaMlvSFp97itIzAjbZ8ZcZur4SbGOscp940Eg7tPKuGBU/olkjwg1kMO35LOLRtx9sOjmbNoWSJxOJeEbBNIGzN7HFgNYGYrgVWVDUbSH4CVwENx00ygi5ltC5wPPCypCCivY7vcTmhJp0saKWnkd999V9kQXULmLl7OJf/9jANveouPp3/PJQN78+K5e7B3r6qrc2SqWcONuPW47Zi7ZDnnP+7zZbnaI9thvIsltSZ+aEvaCZhfmUAknQQMBPY2MwMws2XAsvjzKElfAFsQWhyd0h7eCSgt73nN7E7gToCSkhL/zy4wy1eu5sER07jplYksXr6K43fswrn7bEGrJvWTDm0tfTs2508De/PH/37GP978gv8b0C3pkJyrctkmkPOBZ4HNJb0DtGUDdYv1kXQA8Hugv5ktSdveFphrZqskdSUUy6eY2VxJC2Pieh84Ebgl29/v8o+Z8erns7jq+fFMmR3qHJcM7M0W7ZLpqsrE8bEect1LEynZtBX9NvN6iKvZskogZvaRpP5AD0J30gQzW5HJYyU9AgwA2kiaAVxKGHXVAHg5dkeMiCOu9gAul5TqIjvDzFIr+/yaUItpRKiZeAG9hpjwzUKufH4cb02aTde2Tbj3FzswoEfbxLuqNkQSfz18S8aWLuDsRz5i6JDdK33lu3P5TLG3qGIPks4EHjKz7+PtlsCxZnZ7bsPLrZKSEhs5cmTSYbh1mLt4Ode/PIGH3/+Kpg3qce4+W3DCzpuyUd3Cut51bOl8Drv9XXbq2pr7frGDr6fuCp6kUWZWUnZ7tv+Zv0wlDwAzmwf8MsvncrXc8pWrueutKfS/5jUe+WA6J+y0KW/8dk9O2W2zgkseAH06NOfSQ3rz5sTvuOONL5IOx7kqk20NpI4kpYrdkuoC+VXVdHnPzBg+fhZXDR3Pl7MXs8cWbbnk4F50z+M6R6aO69eFEVPmct1LE9ih2OshrmbKNoEMAx6PU4sYcAbwYs6icjXe598s4MrnxvP25Nls3rYJ9568A3v22DjpsHJGEn85rC+ffT2fsx/5iOeH7L7OFQ+dK1TZ1kDqAL8C9iYU0V8C7jKzSl8LUpW8BpK8OYuWcf3LE3nkg69o1nAjztunO8fvVHh1jkyNK13A4NvfYcfNWnH/yf28HuIK0rpqINmOwlpNmIvqjsoG5mqH5StXc/+7U7n51UksWb6KE3cu5tx9utOicc3u+ezdoYjLDunDxU9/yh1vfMGZe/r1Ia7myHYurF2By4BN43MIMDPrmrvQXE1gZrwyfhZXPT+OqXOWsGePtvzh4F5027jw6xyZOrZf53h9SFg/ZMeurZMOybmcyLYGcjdwHjCKHExh4mqm8TMXcOXz43hn8hy6bdyU+07egQE1qM6RKUn85fAt+fTr+Zz9yGiGnuP1EFczZNvxPN/MXjCzWWY2J/WV08hcwZq9aBkXP/0pB9/8FmNLF/DnQ/vwwjm718rkkdK0QT1uO2475i9dwXmP+fohrmbItgXymqRrgKeIc1VBuEI9J1G5grRs5Sruf3cqtwyfzNIVqzhpl2LO2bvm1zky1btDEZcd2oeLnvqU21+fzFl7dU86JOcqJdsEsmP8nl6VN2CvyoXjCsm8xcsZN3MBY0vnM650AR98OZfS+T+wV8+NufigXnTbuGnSIeadY3YI9ZDrX55ISXErdvJ6iCtgWQ3jLVQ+jDc7ZsbX3y9lbOkCxpUuiN/nUzr/hzX7tG/ekD4dmnPCzpvSf4u2CUab/xYtW8mht7zNomUrvR7iCkJOh/HGJzwY6AM0TG0zs8uzfT6XH1auWs2U2YsZWzqfsV/HZDFzAfOXhrky6wi6tm1KSXEr+nQook+H5vTuUJR306vns6YN6nHb8dsx+LZ3OO+xMdx3cj/q+vUhrgBlO4z3H0BjYE/gLsJU7h/kMC5XDZYuX8X4bxasaVmMK53P598sZNnK1QA0qFeHnps046At29OnQxG9OxTRc5NmNK6f9XmHi3q1L+LPh/bhwqc+5fbXJnP23l4PcYUn20+CXcxsK0mfmNmfJV1HKKi7PDV38fI1tYqxpaFu8eXsxaQGAzVvtBF9OhRxwk6b0qdjaFl0bdOEejX0CvF8cHSsh9zwSqiH7Ly510NcYck2gSyN35dI6gDMATbLTUiuMsyMGfOWrqlThCL3Amam1Ss6tmhEr/ZFDNyqw5qWRccWjfJ+vY2aRhJXHbYln3w9nyGPjmbokN1p28zrIa5wZJtAnpPUArgG+IgwAuuuXAXlMrNi1Wq++G4RY79esNZoqAU/rARCvWLztk3ZcbNWa2oVvdsX0dLrFXmjSbw+JFUPuf8Ur4e4wlHpUViSGgANzSyjNdEl3UNY+3yWmfWN21oBjwHFwFTgZ3GNESRdBJxKuOJ9iJkNi9u358cVCYcC59gGDqaQR2EtWb6S8TMXpHVBLWDCtwtZHusVDTeqQ89NQmsiVdzu0a4ZjerXTThyl4lHP/iKC5/6lPP33YIhXg9xeSYno7AkHb6e+zCzTOog9wG3Ag+kbbsQGG5mV0u6MN7+vaTewDGE0V4dgFckbRFn/b0DOB0YQUggB1BDlrWds2jZmiSRall8OXsxqfTYonGoV/xil+LQBdW+iM28XlHQjt6hM+9/OZcbX5lISXFLdtm8TdIhObdBFe3COmQ99xkZFNLN7E1JxWU2DyKskw5wP/A68Pu4/VEzWwZ8KWky0E/SVKDIzN4DkPQAMJgCSyBmxvS5S0PX08wfi9vfLlhzcT8dWzSiT4ciDt26A306NKdPhyLaN2/o9YoaRhJXDu7LxzO+55xHx3g9xBWECiUQMzu5iuJoZ2Yz4++YKSk1aVJHQgsjZUbctiL+XHZ7wbh22ATuf3cqC5eFekXdOqJb26bssnmbNYXt3u2LfBqQWqRJg3rcfvx2DLr1Hc59bDQPnLKj10NcXsv3CwnL+++x9Wz/6RNIpxO6uujSpUvuIqukTVs3ZtC2oVXRu30RPTZpRsONvF5R2/XcpIjLB/Xh909+yq2vTuacfbwe4vJXvlxI+K2k9rH10R6YFbfPADqn7dcJKI3bO5Wz/SfM7E7gTghF9ErEmFNHlXTmqJLOG97R1To/K+nM+1PmcuPwiexQ3JJdunk9xOWnbKuuu5jZicA8M/szsDNrf9BX1LPASfHnk4Bn0rYfI6mBpM2A7sAHsbtroaSdFIoBJ6Y9xrmCJokrBvela5smDHl0DLMW/rDhBzmXgGwTSNkLCVeQ4YWEkh4B3gN6SJoh6VTgamBfSZOAfeNtzGws8DgwDngRODNt3fVfE1o/k4EvKLACunPrE+oh27No2QrOfXQMq3z9EJeHcnkh4b8yeaCZHbuOu/Zex/5XAVeVs30k0DeT3+lcIeqxSTMuP7Qvv3vyE255dRLn7rNF0iE5t5asEoiZXRF/fFLSc1TgQkLnXOaOKunEiC/ncNPwSfQrbuX1EJdXKtyFJWlTSW3izzsBZxOK6c65HEtdH7J526ZeD3F5p0IJRNIlwKvACElXAjcCbYBzJN2Y8+icczSuH+bLWrRsBec84vUQlz8q2gI5FuhFWMp2CLCXmV1IKHzvk+PYnHNRj02acfmgvrw3ZQ43D5+UdDjOARVPID+Y2XIz+x74wsyWAJjZSmB5roNzzv3oqO07cfh2Hbn51Um8M3l20uE4V+EE0kLS4ZKOAIriz6nbzasgPudclF4POcfrIS4PVDSBvEGYUHEg8Gb8Of22c64KNa4f5svyeojLB/kymaJzLkNbtGvGFYP68tsnPuGm4ZM4f1+/PsQlwxeQcK4AHVXSmSO268Qtr07i7UleD3HJ8ATiXIG6YnAfurVtyrmPjWbWAq+HuOqXVQKJy9hucJtzruqk6iGLl61iyKOjvR7iql22LZD3MtzmnKtC3ds144rBfRkxZS43vTIx6XBcLVPRNdE3Iaz810jStvy4sFMRYX0Q51w1O3L7ToyYModbXpvMDpu1YvfubZMOydUSFZ1McX/gF4QFnK7jxwSyALg4d2E55yri8kF9+Hj695z76BiGnrM77YoabvhBzlWSzCrWbyqpDnCsmT1UNSFVnZKSEhs5cmTSYThXJSZ9u5BDb32HrTo156HTdqRe3Zo3RmbxspWMn7mAsaULGFe6gLEz59OycX1+f0BP+nb0a5mriqRRZlbyk+0VTSDxyd40sz1yElk18gTiaronR83ggv98zNl7deOC/XokHU6lfLdwGWNL5zMuLWFMnbOY1EdWqyb16d2+iPEzFzB3yXKO3K4Tv92/Bxt76yvn1pVAsl1Q6mVJvwEeAxanNprZ3CyfD0k94vOldAX+BLQAfgl8F7dfbGZD42MuAk4FVgFDzGxYtr/fuZrgiFgPufW1yexQ3Io9tsj/esjq1cb0eUsYW7qAsaXz1ySLWQuXrdmnc6tG9GnfnMO27UifDkX06dCcdkUNkMSCH1Zw26uTueedL3n+05mcuWc3Tt1tMxpuVDfBo6odsm2BfFnOZjOzrpUPCSTVBb4GdgROBhaZ2bVl9ukNPAL0AzoArwBbpC15+xPeAnG1wdLlqxh029vMWbQ87+ohy1euZtKshWuSxLjSBYybuYBFy1YCULeO6L5xU3rHJNG7fRG9OxTRvNFGG3zuqbMX89cXxjNs7Ld0bNGIiw/qxUFbboKkDT7WrV9OWyBmltH655WwN2G232nrefEHAY+a2TLgS0mTCcnEhxO7Wq1R/brcfvx2HHLLOwx5ZHRi9ZCFP6xg/MyFjIutirGlC5g0ayErVoWT1sb169KrfdFarYru7Zpm3XIobtOEf55QwrtfzOaK58Zz5sMfsUNxS/40sA9bdvL6SFXItgsLSX2B3sCa0xszeyAXQQHHEFoXKWdJOhEYCVxgZvMIw4lHpO0zI25zrtbrtnEzrhzclwv+8zE3vjKJ3+xftfWQWQt/WKtVMbZ0PlPnLFlzf+sm9endoYg9tuhKnw6hVVHcugl16+S+dbDL5m147uzd+M/I6Vz70gQOve1tjoj1kXxqjdUE2XZhXQoMICSQocCBwNtmdmSlA5LqA6VAHzP7VlI7YDZgwBVAezM7RdJtwHtm9u/4uLuBoWb2ZJnnOx04HaBLly7bT5s2rbIhOlcwfvfEx/xn1AzuO7kf/XNQD1m92pg2d0kobsdWxdjSBcxe9GO9okurxiFJtC+iT8fQsti4WYNEupIW/rCCW1+bzL1vT6VeXfF/AzbntN27en2kgnI9CutTYGtgtJltHT/k7zKzQ3IQ6CDgTDPbr5z7ioHnzKxvLKBjZn+N9w0DLjOzdXZheQ3E1TZLl69i8G3v8N2iZQwdsjubNM/8DHzZylVM+nbRmhbF2NIFjJ+5gMXLQ5mxXh3RbeOm9OnQPHZBFdGrQxFFDTdcr6hu0+Ys5q9DP+fFsd/QsUUjLjywJwO3au/1kQzlOoF8YGb9JI0C9gQWAp+ZWZ8cBPooMMzM7o2325vZzPjzecCOZnaMpD7Aw/xYRB8OdPciunNrmzxrEYfe+jZ9Ozbn4XXUQxb8sILxsTWRGjY7Oa1e0STWK1LdT6l6RYN6hXUm/94Xc7j8uXGMn7mAkk1bcsnA3mzduUXSYeW9XCeQ2wlXnh8DXAAsAsZUdr0QSY2B6UBXM5sftz0IbEPowpoK/CotofwBOAVYCZxrZi+s7/k9gbja6unRMzjvsY85c8/NOXHn4p90QX0198d6RZum9emd1qro3T7UK+pUQb0iCatWG0+Mms41wyYwe9FyDt+uI78/oKfXR9YjpwmkzBMXA0Vm9kmlnqgaeAJxtdnvn/iEx0ZOX2vbpq3T6hUxadSWC/EW/rCC2177gnve/pK6dUJ95Jd7eH2kPLlugQw3s703tC3feAJxtdnS5au47bXJtG5anz4dmtOzfbO8rFdUt6/mLOGvL4znhc++oUPzhlx4UC8O8frIWnKSQCQ1JMy6+xphFFb6bLwvmFmvyodadTyBOOfWZcSUOVz+v3GMm7mA7WN9ZBuvjwDrTiAVvbroV8AooGf8nvp6BritskE651xSduramv+dvRt/P2Irps1ZwuDb3uH8x8bwzXxf7XFdsu3COtvMbqmCeKqUt0Ccc5lYtGwlt702mbvfCvWRXw/YnF/u3pVG9WtnfSQnLRBJO0jaJJU8JJ0o6RlJN0tqlatgnXMuSU0b1OP3B/Rk+AX92bNnW65/eSJ7X/c6z4z5msoOPKpJKtqF9U9gOYCkPYCrgQeA+cCduQ3NOeeS1blVY24/fnseO30nWjapzzmPjuGIO95l9Ffzkg4tL1Q0gdRNm7L9aOBOM3vSzC4BuuU2NOecyw87dm3Ns2ftxt+P3Irp85Zy2O3vct5jY5g5f2nSoSWqwglEUmoCxr2BV9Puy3piRuecy3d164iflXTmtd8M4Mw9N+f5T2ey17VvcOMrE1m6fJ0TYNRoFU0gjwBvSHoGWAq8BSCpG6EbyznnarSmDerx2/17Mvz8/uzVa2NufGUSe133Ov8d/TWrV9eu+kg2a6LvBLQHXjKzxXHbFkBTM/so9yHmjo/Ccs7l2gdfzuXy58by2dcL2KZzC/50SG+269Iy6bByqsqmMikknkCcc1Vh9WrjyY9m8PdhE/hu4TIGb9OB3x3Qkw4tGiUdWk7k6kJC55xzZdSpI44q6czrvxnAWXt2Y+hn37DXda9zw8sTWbJ8ZdLhVRlPIM45lyNNGtTjN/v3YPj5/dmnVztuGj6Jva59g6dHz6iR9RFPIM45l2OdWzXm1uO24z9n7EzbZg0477GPOfyOd/mohl0/4gnEOeeqyA7FrXjmzF259qitKf1+KYff/i7nPDqa0u9rxvUjnkCcc64K1akjjty+E6/9ZgBn79WNF2N95PoaUB/JqwQiaaqkTyWNkTQybmsl6WVJk+L3lmn7XyRpsqQJkvZPLnLnnFu/Jg3qccF+PRh+QX/27b0JNw+fxJ7Xvs5THxVufSSvEki0p5ltkzZk7EJguJl1J6x7fiGApN6EJXX7AAcAt0uqnVNlOucKRqeWjbnl2G154oydaVfUkPMf/5jDbn+HUdPmbvjBeSYfE0hZg4D748/3A4PTtj9qZsvM7EtgMtCv+sNzzrmKKyluxX//b1euO2prvlnwA0fc8R5DHhnN1wVUH8m3BGLAS5JGSTo9bmtnZjMB4veN4/aOQPoCzzPiNuecKwh16ogjtu/EqxcMYMhe3Rg29hv2uvZ1ho//NunQMpJvEyDuamalkjYGXpb0+Xr2LW/B4p90JMZEdDpAly5dchOlc87lUJMG9Th/vx4c3a8LN70ykW0LZCqUvGqBmFlp/D4LeJrQJfWtpPYA8fusuPsMoHPawzsBpeU8551mVmJmJW3btq3K8J1zrlI6tmjE34/cmlZN6icdSkbyJoFIaiKpWepnYD/gM+BZ4KS420mE9deJ24+R1EDSZkB34IPqjdo552qvfOrCagc8LQlCXA+b2YuSPgQel3Qq8BVwFICZjZX0ODAOWAmcaWa1c1J+55xLQK2ajVfSd8C0pONI0waYnXQQOVbTjqmmHQ/UvGOqaccD+XdMm5rZT2oAtSqB5BtJI8ubIrmQ1bRjqmnHAzXvmGra8UDhHFPe1ECcc84VFk8gzjnnsuIJJFl3Jh1AFahpx1TTjgdq3jHVtOOBAjkmr4E455zLirdAnHPOZcUTiHPOuax4AnHOOZcVTyDOOeey4gnEOedcVjyBOOecy4onEOecc1nxBOKccy4rnkCcc85lxROIc865rHgCcc45lxVPIM4557LiCcQ551xWPIE455zLSr2kA6hObdq0seLi4qTDcM65gjJq1KjZ5a2JXqsSSHFxMSNHjkw6DOecKyiSppW33buwnHPOZSXRBCLpAEkTJE2WdGE59x8v6ZP49a6krdPumyrpU0ljJHmzwjnnqlliXViS6gK3AfsCM4APJT1rZuPSdvsS6G9m8yQdSFgneMe0+/c0s9nVFrRzzrk1kqyB9AMmm9kUAEmPAoOANQnEzN5N238E0KlaI3QFRX9W0iFkxC61pENwLieS7MLqCExPuz0jbluXU4EX0m4b8JKkUZJOr4L4nHPOrUeSLZDyThfLPTWTtCchgeyWtnlXMyuVtDHwsqTPzezNch57OnA6QJcuXSoftXPOOSDZFsgMoHPa7U5AadmdJG0F3AUMMrM5qe1mVhq/zwKeJnSJ/YSZ3WlmJWZW0rbtT4YxO+ecy1KSCeRDoLukzSTVB44Bnk3fQVIX4CngBDObmLa9iaRmqZ+B/YDPqi1y55xzyXVhmdlKSWcBw4C6wD1mNlbSGfH+fwB/AloDt0sCWGlmJUA74Om4rR7wsJm9mMBhOOdcrZXolehmNhQYWmbbP9J+Pg04rZzHTQG2LrvdOedc9fEr0Z1zzmXFE4hzzrmseAJxzjmXFU8gzjnnsuIJxDnnXFY8gTjnnMuKJxDnnHNZ8QTinHMuK7VqSVvnXLJ8yv2axVsgzjnnsuIJxDnnXFY8gTjnnMuK10BqMe+Pds5VhrdAnHPOZcUTiHPOuax4AnHOOZcVTyDOOeey4gnEOedcVjyBOOecy4onEOecc1nxBOKccy4riSYQSQdImiBpsqQLy7lfkm6O938iabtMH+ucc65qJZZAJNUFbgMOBHoDx0rqXWa3A4Hu8et04I4KPNY551wVSnIqk37AZDObAiDpUWAQMC5tn0HAA2ZmwAhJLSS1B4ozeOxPTJgwgQEDBuT6OArX1KQDyMyA1wZktuPUqowidzI+nppoatIBZKZWv0YVkHECkdQR2DT9MWb2ZiV+d0dgetrtGcCOGezTMcPHAiDpdELrhQYNGlQi3Jqnf3H/pEPIqZp2PAC88UbSEWSmf2Z/e3+NEpTha1QRGSUQSX8Djiac4a+Kmw2oTAIpbya/srPmrWufTB4bNprdCdwJUFJSYq+//noFQnQuYSqMCS+pzf9XteA10jqOMdMWyGCgh5ktyzqCn5oBdE673QkozXCf+hk81jnnXBXKtIg+Bdgox7/7Q6C7pM0k1QeOAZ4ts8+zwIlxNNZOwHwzm5nhY51zzlWhTFsgS4AxkoYDa1ohZjYk219sZislnQUMA+oC95jZWElnxPv/AQwFDgImxxhOXt9js43FOedcxSkMcNrATtJJ5W03s/tzHlEVKikpsZEjRyYdhnOZK5T+9Qw+R2qsWvAaSRplZiVlt2+wBRKvuTjBzPbJ+rc755yrcTZYAzGzVcASSc2rIR7nnHMFItMayA/Ap5JeBhanNlamBuKcc66wZZpAno9fzjnnHJBhAim0Yrlzzrmql+mV6F9SzpXeZtY15xE555wrCJl2YaUP32oIHAW0yn04zjnnCkVGV6Kb2Zy0r6/N7EZgr6oNzTnnXD7LtAtru7SbdQgtkmZVEpFzzrmCkGkX1nVpP68EvgR+lvtwnHPOFYpME8ipqcWbUiRtVgXxOOecKxCZzsb7RIbbnHPO1RLrbYFI6gn0AZpLOjztriLCaCznnHO11Ia6sHoAA4EWwCFp2xcCv6yimJxzzhWA9SYQM3sGeEbSzmb2XjXF5JxzrgBkWgOZI2m4pM8AJG0l6Y9VGJdzzrk8l2kC+RdwEbACwMw+ISwj65xzrpbKNIE0NrMPymxbmetgnHPOFY5ME8hsSZsTJ1SUdCQws8qics45l/cyvZDwTOBOoKekrwlXoh9fZVE555zLe5lOpjglroneFugJDAB2y/aXSmol6WVJk+L3luXs01nSa5LGSxor6Zy0+y6T9LWkMfHroGxjcc45l531JhBJRZIuknSrpH2BJcBJwGQqNxfWhcBwM+sODI+3y1oJXGBmvYCdgDMl9U67/wYz2yZ+Da1ELM4557KwoRbIg4SLCT8lXDj4EmEtkMFmNqgSv3cQkFrl8H5gcNkdzGymmX0Uf14IjAc6VuJ3Ouecy6EN1UC6mtmWAJLuAmYDXeIHemW0M7OZEBKFpI3Xt7OkYmBb4P20zWdJOhEYSWipzFvHY08HTgfo0qVLJcN2zjmXsqEWyIrUD2a2Cvgy0+Qh6RVJn5XzVaGWi6SmwJPAuWa2IG6+A9gc2IYwGuy68h8NZnanmZWYWUnbtm0r8qudc86tx4ZaIFtLSn1oC2gUbwswMyta1wNj0b1ckr6V1D62PtoDs9ax30aE5PGQmT2V9tzfpu3zL+C5DRyHc865HFtvC8TM6ppZUfxqZmb10n5eZ/LIwLOEYjzx+zNld5Ak4G5gvJldX+a+9mk3DwM+q0QszjnnspDphYS5djWwr6RJwL7xNpI6SEqNqNoVOAHYq5zhun+X9KmkT4A9gfOqOX7nnKv1Mr2QMKfMbA6wdznbS4GD4s9vE7rKynv8CVUaoHPOuQ1KqgXinHOuwHkCcc45lxVPIM4557LiCcQ551xWPIE455zLiicQ55xzWfEE4pxzLiueQJxzzmXFE4hzzrmsJHIlunPO1RhmSUeQGG+BOOecy4onEOecc1nxBOKccy4rnkCcc85lxROIc865rHgCcc45lxVPIM4557LiCcQ551xWPIE455zLiicQ55xzWUkkgUhqJellSZPi95br2G+qpE8ljZE0sqKPd845V3WSaoFcCAw3s+7A8Hh7XfY0s23MrCTLxzvnnKsCSSWQQcD98ef7gcHV/HjnnHOVlNRsvO3MbCaAmc2UtPE69jPgJUkG/NPM7qzg45F0OnA6QJcuXXJ2AM5Vi1o806vLf1WWQCS9AmxSzl1/qMDT7GpmpTFBvCzpczN7syJxxKRzJ0BJSYn/NzrnXI5UWQIxs33WdZ+kbyW1j62H9sCsdTxHafw+S9LTQD/gTSCjxzvnnKs6SdVAngVOij+fBDxTdgdJTSQ1S/0M7Ad8lunjnXPOVa2kEsjVwL6SJgH7xttI6iBpaNynHfC2pI+BD4DnzezF9T3eOedc9ZHVoiKdpO+AaUnHkaYNMDvpIHKsph1TTTseqHnHVNOOB/LvmDY1s7ZlN9aqBJJvJI0sc31Lwatpx1TTjgdq3jHVtOOBwjkmn8rEOedcVjyBOOecy4onkGTdueFdCk5NO6aadjxQ846pph0PFMgxeQ3EOedcVrwF4pxzLiueQJxzzmXFE4hLhCQlHYNbP0l1JdWLP9eIzwpJ7ST1Sc1yUcjy4TXxGkiek9QV2AmYZ2YvJB1PLkgaAAwERhFmGFiQaEA5IGkPYCugFBhmZosTDqlSJPUGfgc0AG4zs7cTDqnSJPUF7gYWAROB88zsh2Sjyo6k7sCvgAnAc6nZyatb4hnMrZukLYChwDbA/yQdnWxElSdpf+AuwlW2fwR2TzaiypN0KHAz0BU4Atgh2YgqR1I34F7gHeBd4HJJG8X7CvIzQ1Jr4HbCtEf7AH3iV8GJyeN5YCFwKnB0mfurrXWf1HogbgMkNQCuIayDcoOkD4HmkrYws4kJh5eVOCnmCYQzv/9Jmg+USJoHTDez6clGWHHxmE4BTjWzUZL+BvSQNAP41swWJhthVvYG3jOzf8WungHAzyVNNLN3JMkKr+uiCbAEeNvMTFJj4HeSvgFeNbOCmJA1JvBTgfvM7C9x2YzjJO0JfGlmU6vztSnIs4nawMyWEbp4kNQRuAI4FHhF0oWFeCYYu3WmASdL2oFwNtgZuBw4M3abFBoDWgHbxnVrDgcOAi4F7pDUPMngsjQHaCHpeOBFoC6wMfCgpP0LMHkAfAOMIbTk3wc+B84DvgYOWN+idPnEzFYDk4EDJW1H6JLbBLgYGCKpf3XGU3AfQrVBWnKYDvQGHgJGmtmhwP6Es/ijEgqvwso0qR8CvgJ+C9xhZqcBZwGbAT0TCK9SzGwJcBlwJvAY8KiZDYrbfiDUrwrNMGAKsAWw3MwGm9nfgD8DF8XWcd6T1FJSGwAzWw78jfA6jQeujOsN3Ug4zrxerrTM/9Aw4A1gCPCpmR0BHA9sRDV3n3oCySOSNpfUO55lYGb3Ej5cbwc+llTHzMYT+qdbJRhqxmLB/GeSWgCY2TgzOxf4L9BQ0kZm9jkhqfSIj8nrEVqSdpR0WOq2mb0K7Aq8BIyN274gdBHn/ZmtpL6SHkuduJjZQjO7nFDX+TAtYUwhLN6W9y0QSX2Aj4HTJG0KYGbzzGwUYXmIwyW1AjoS/peWJhbsBsRa6Clx8TzMbLqZ/RH4J7BKUkszm0XoseguaaPq+h/yGkiekLQ5oVm9QNJeZvYxhK6sOA39vsAASY0Ia7z/MrloMyNpd+BVwgfrSknDzez7ePeHhLOlP0v6ChgEHAyQz10kcRDAk8CLkpqZ2QMQWiKS3gYujf+7K4G+hK7HvCWpA6FVuDEhWexgZqsl1SXUDLYGbo1r7xwLXBbP5vNW/B85B/iI0EV6iKRn0mpsE4AjgaeB5sDlZjY2kWA3QFIXwv/Kl8BqSS+mjbhKnXhdImkkYbnwM81sRbXFl8f/q7VG/Gc9nfChsxFwNnCsmY1J2+dawpu9F/AXMxtazlPljXj9wEDCMNC6wAHA/4CXzGx+PKs9lNDF0w74a77+E6eTdC7QjHB2O5BQlH0g3tcEOITwWs4HLjWzTxIKNSOStgJ2MbN/SHqIMDJpu1QrOI5e+hXhdXzHzF7K9yK6pIZALzMbHVvApwEjgP+Z2bS4TxugGFhqZmPz8ZhiK2I3QtfuDEI31RuE4/gm7jMQ2BHoTiisv7iOp6uaGPPsb1ZrxTPB1Wb2jaQLCG/648xsdJn92pjZ7Hx8w5cVR/CsNLOlkk4mjOYZCrxsZnPT9mtYKOPxY2LcCKhPSIC7EUYs3Ze2TyNCQ6pQjmnj2AVCWhLpZ2bLJTWOdZ7Uvnn7vosDFpab2dIy2wcQRi59QOgO7g58lX5c+Sq+lxqa2TxJBxFagW8BL5rZV2n7NU7ieDyBJCj2aXYBJqR17aTuSyWRPQln6P3N7OZYB1ld7cFmSFI/wgfQB8A3ZjYn7b5TgT0INZwdgKZmdmk+fygBxBFjdYFl6Qk99qEPJCSRp4EWwPf5fsFnHO12IGFo67WE41qVdv/DhLPzSwgfvOenznjzlcJFgjcSahlvA6PM7JW0+/cidFtBGIRysJm9Wd1xZkJSMaHL+iOgNP0iQUkHE5LIU4RW4X5mdrKkuumvYbUxM/9K4Iswumo6YeTOKEKy6F5mn5MIF9zNAg5MOuYMjmkgodB6F/AAYYjhluXs83E89m2SjjmDYzoQmBSP6X/AzWXub0XonvsQWAD0TTrmDRxPJ0K/+QXA48CjhNpTszL7vQusBg5POuYMjqlJ/B86idAl+mvCII2jyux3IeEq9EOSjnk9x9KDUO/4V/xsuBvYv8w+2wFvxs+GnyUab9J/sNr4RRj9diXw63j7MODvhCuzu6ft1x1YRjhbSjzuDI7rKkLtJpUgzwOeTf9QJZwFLgB6Jx3vBo5FhK6qp9OOqW38oLq3zL4XEfqo8/qYYqwDgYfTbp9NWHtiILBR3NaNUGg+NPW3SDruDRxTQ+DBVBKMSf1w4BnggLitWXwv/izt9c274wJOBq6IP3cntJaGpicRQk1kJTAw6dfHh/EmwEIXVCPCmQRm9jThzV0E7KOgCeGf+nAzez7fh7ZGDYFdIAzXJYzueY1wkWDLtJE9O8f785YFKwgJY1Xc9h2hYLmlpOvTdq9H+LDN62OKPgE6SNoNwMxuIbQIjyV8yELoivuFmT1bCO87C7WmesA98fZcwui//wI7xbrVUuAkM3s8dUwWP33zTH3C1EWY2SRCq/dhwkwAm8d9WhB6JJ5LuvvXE0g10tpXj98YNoXrCSxMVvcmcCLQ1sJV23eY2fPVHmgFSGoYkx2E/vQdJJ0NYKEw+yrhDd/SQh/tC5bno63iKJ6Ur4CL42gkzGwlYchxsaSecdsVZvZR9UeaGUmtJTWJAzC+Iozk2SleX4CZ3Ua4tuOyeHukmb0Xf7Z8/KCV1FXSYQpXywP8gjBU/FoACzXF0cDOhFrbSjObF+/Lq2OS1CAWyzGzfwL1JF0Tb39PqOmsIHQ/ArxvZi8nEWtZnkCqSfxnvUfS5ZJOtDAm/SNgN0mDAczsOcIsoSXx9ppRPPn0hk+JBb3HgKGSTrFQ7Ps9sK+kIQAWrmepA2wfb+fdcaSTdCBws6S7JW1lYYjuq8D7+vGq5q8J/9B5f0W2pF6Ebrh/ATfE248Qrr4+RNKOcdfHCZPz5b14DP8l/J9cLekWC1P/XAG0k/RwbO3WIwx9L0os2A1Ie30ek3R13HwO0Cl128ymAt8SkuFa/0OJ/z8l1XdWm74IfZYfE+arOYHQ0tgZaE2YjuA24CZgL2AmhVFc3h/4lDDx3r6EuYb2ifftRhh3fzshoUwGipOOOYNjOoBwcdZ+wK2ED9p68b5rgZGEM93fEabD2DTpmDdwPMWEK+OPJhRnfwP8Id63M/AXQv/6PwnT0A9KOuYMjqll/P85Le32h8C2hLpGC8LAgCeBccCRSce8nmPZgjA/16nx9ZlCuBAQwkWo/yPUcU4hDDrpn3TMZb98GG8Vi90hVwOfW7hYqy5heOQUM3tAUkvCZGi/IxTMh1moieSt2Ny+mDBU8r9x29mEaz7uiLfbEC5AA3jWzD5NItZMSWpKGPHypJk9Hre9ADxjZv+It48nXLFdQrjw8bOk4s2EpGOBzczsL/H2IcDZZrZfvN2ScDy7AOPM7P2k+9Q3RFJbwgfpE5LqmdlKSY8T1ix5I22/1oTk/20+HlP8HPgVUN/Mbozb9icMy70g3hZh5NhywudH3nVnewKpBpJ2BaaZ2Yx4+yxgNzM7psx+DSxMXZJ3b/iyFOYamg4sNDOLXVa7lD2mQiKpB2F21h/iB9NVwGwzu6HMfsmMuc+CpM0tzMuVSur3EQZmLJfUxApw4StJrW3t64uuJ0zJ/pyk7haKz3kvFsWXWLzOQ9I+wF+B3a2ci1Dz8XPBayBVKG20xzup5BFNIzS3UxPzHRr3Wxa/59WbpDxmNtbMFqTFOoHYhy7pBEmDkosua1+Y2SILhXIIr1MTCItGSdo9vqZ5eyFnStp774u0zQZ0AFYorKB4o6TmhTDSKl0qeaQNSmkFNIj1nGGxrpD3zOwLW3slwUmEC1F/iO+1M8vsn3efC55AqlDZFzztH3UG8GV8o/+TMO13ofsGmK2wauKFhLpHXku9Hmmvy6oyt+sCdSQdTrhOZ4ZF1R5sBa0jxmWERD+YMArwKTObXwjHUx77cUaGcYTri64DhliYsboQzQGmK0y9cjuhNZzXvAuriqyvuamwEMy7wBfA7/KxbzNdeceiMlOqSNqFMNxwDGEOr8+rN8qKST8mrWOVx9iKup0wMu4sy/PhxykKU698X15XjqT3CK2QMyzPp1xJJ2lbwjxXY8t2IUr6FSEhHmZmL+ZjV095yrwH6xCmLJpCOBk7w8yG5fuxeAskx9LOXpusYzuErp5SQkEz35NHnbQ3eUdJm8BaZ38pUwnrNOd98oAfz9AlnURYObBJOV05XxFah2cWUPLoThi0sa5uqUmEZFgwySM6BrgeoJz600fAHvmcPNJau+304zUfph/XYFlNmJrkeWLySO2TUMgZ8RZIFVCYNfMiwjoY71mc1K28s958fcMDSNoeaGxmb8Ui+S8Icwl9amZnxn3WtEQktbAyk0Lmszi44RLCVdfflHNmWx9oZGbzEwuyAuJZ+rOEqTDulFTf4todZd9n+fy+g/LjJYyS+5eZvbeu+PP5uBSmXr+OcCHnN2b2p7g9/X9oYzOblc/Hkc5bIDkWu6dOI8wv1Bg4SNKRsOaMY6P488TUtqRizUB/4EpJPydM4XEQYSndrSTdDuHMKe0s6vukAs1Ees1DYdjudoSZgw+HcGab3goxs+WFkjwALMwUPAk4P95erjBc9Cfvszx/36X+V3aV9HNJO8d4ZxGuMVpn/Pl6XAoj/A4GziWMhOsm6TpY8z+U+lyYFb/n5XGU5QkkhyR1JlzR+5GZPQjcQujT3EXSMQBWjauFZSttBM/1hAuyziJMLLjQzL4lXES4Rywul9edlXfKnNE1sDDa6hbCpJbb68fZAKzQRiVJ2lrSAbEFuBfwhaTXYE1SLNSVR5sRZjD4k6Q/Ae8DpyoMIS8YktoRZjNoFLsOPyS87zaRdBsUxudCeTyB5JCF6UmeIUwe2N3MSoH/EEZT7JqqH+Szsk1nM7uZ0PdcDGwnqamFhWteIswImvfKdB0OAe5VmKqkv4W5hz4gtBSPhsI4+0trTe1FmGzvfOB2SWea2cHAPEkjYM38XXkv7ZhKYnfc52Z2HmGix7aEZNIpfk9djJfXJHWIJ10XALtL2j0mi88Jtap2KpBhx+WyPLgcvhC/SJsOmjANwdZp911COMvoEW9vQp5Pe1HO8f2SMLfQHwhngqcRksZf432fA1skHWcFj+lMwuzAmxHmUppKXBuCMP/QzYSJ9xKPNcPj2ZYwFUnqfbZvPIbUFOZvADsmHWeGx1I3fj+YMP3K+YTJEFNTltcjzFR7LmHYbuOkY87gmNoTFk87Pd4+gTC8fbd4uw7QJOk4K/NVqE3bxFl8ByhMKHgt8FY8axpsZldIWg78V9JgM5uQZKyZUJwWIv48hLBc6+WE4ZFmZn+RZIQ1S+4hrE8wLal4M6FwZe/OZnZF3LQRod5xMuFiwN8DN0labWY3xS6gRQmFWyFxJM+OwD6EudYmAO8A/QiLYL1oZv2TizAzklqZ2VwLXW29CTMCDySsXLkRcIvCDA1PxofcqDBkvDthfrm8ZWYzJb1FmPl4uZndF/+HnpD0MwsrIhbcTABrSTqDFdoX0JGwqD2EhXfGEZalHUx4M7xNnDiQ8GG7a9IxZ3BMhxBaG3UJZ0XXxJ/PJ5zhNuDHxYZOIM9bU/y4GNQ4wpDpy9O2dwOGA63jtuGESSHz+kyQ0Mo9Buictq0FoYX4MmEaGQhn8M8QWo11k457A8dUTFh97+p4uwnQi1AoHwM0JUwA+T1xZUTCJIOfAl2Sjr/MsaR/LmxGuL4rdd/xhJOuk+Ltk4ABSceciy+vgVSQham8W0l6iVDb2J+w+t7FQJu47RVJm5nZlWb2TnLRbpikfQnJ4237cQhre8KMp7sQZmhdRiheHmJmD1qetzwsWAH8lrB2dBdJN1v4750NfAe0V7gG5EPCLMJ5eyYYawOnE1bd+5ukGxQm6VxEuEL+ReA/CtN/nwncZWYLLf/n61pJOGEZIOkaM1ts4Sry3sBDFlqDXxEGcsyLj/ka2NPCuiZ5I34utJb0OtCZUC+8IN73EKHr6hKFZQ/uN7PXC22wRnk8gVRA2nDVXQldIC9aKJx3J8yiu5QwCmsRYR2CvBYLsE8TlvkcJqmY0CVyA2Gq+efMbIWkXxBqBHm/4l4c8ZIyjbAWxFPAEkk3WhhqPIXQffUH4EELRc68FRPfMMIopEsIZ+5XA38mnLXfRLi+YAfgUTP7n9ZevCwvWZgf7hbCe9BSI5IISx7vJuk8wvHebGavxcEQ88xsdkIhlyvtc2EXQuzXEZYD6C3pt3G35wh1wxGpx8XXtaDl/Zssn9ja1zwcACyV9AphVtrWkv5MGPJ6qpmNSS7SjM0mLK1bHI/rEaCbmY0idF9dLOlBwlntEbb2xHx5J9Y8PpN0taROhGLsf4DjiEsGS7rMzC4m1EF2tgK5wtzMXiK8XseZ2WGEq6/PJSSWswjzq/0bOE1SieXp0GpJm6VGu0UfE+pSLwKrJF1tZo8Skn4r4GILi5Ll7Qdumc+FQwlr+txAKKDvJekZwv/W9VYYyx5nzK9Ez4LWvnL0acKwwt8RFul5zeIaGYVAYd6klwgTCZ5tZo+k3deGcJKxOt/O+sojaSvgPcL6138gnJE/Sbhg8D5Ca+QKYGJMIgUh9X6T1I9Qr3qCsN789YTEsSvhLH4KYfGoVMs4ryhc2T+RUDO8mjB8+nVC0bwF8BahpTvfwvDd1OMK4qrsMp8LzxLeb4cTRsfNtFA0r1E8gWSpzJvlScJFQgeVva8QSNqSUPP4tZk9GsfX17ECvLgpjql/nTAJ4guED9k+wIVm9s94EdpcW3sa7YIgaWNC4tgNONfCNSwobV0P5flaJZK2Jgyh/oowwu8CQiuksZmdHE8CzgOuKcSz9TKfCy8QFlk7pOx9NYV3YWWpTLP1CGCZflwHvKDeJBZWC9yPMKngGWa2qhCTB0Aswh5E+BDqaGa7Ec5wP4r3jy3E5AFrprn4A+ED91lY86G0OG2fvE0eALE7ahBhNFUzwqg+CIMatiCMsBpSiMkDfvK5cCCwUmG1zoL7XMiEXwdSCak3S3xjfED4hyhIZvZhrCF8KGmZmd2bdEzZMrNRkvYDXorXGdyddEw5NJpQ29ld0hOF+KFkZp9IOgB4hdDqPVNSC2BR7KpamGiAlVTO50JR0jFVFU8glRTfLA0IyePfScdTGfGDd3tgSdKxVJaZfSBpb0JCXF3ICTFdHBX3T8J63wWXPFLSTliel9TGwpQ5NUZN+lxYH6+B5Ej6ldwuf8TZAZZYAcwGUBspLEP7CqFONaOQk2J5avrngicQ51yiJBWZ2YKk43AV50V051zSFsJPVu10BcBbIM4557LiLRDnnHNZ8QTinHMuK55AnHPOZcUTiHM5JMniBJSp2/UkfSfpuSyfr4Wk/0u7PSDb53Iu1zyBOJdbi4G+ccVACBPpfV2J52sB/N+GdnIuCZ5AnMu9FwgrAwIcS5jKGwhLuEr6r6RPJI2Ikwci6TJJ90h6XdKU1LxqhFlrN5c0RtI1cVtTSU9I+lzSQz781SXFE4hzufcocExcNXArwkJQKX8GRpvZVoRVLB9Iu68nYYXLfsClkjYCLgS+MLNtzCy1ONG2hLVAegNdCdO5O1ftPIE4l2Nm9glh1cBjCWvKp9uNsDQtZvYqYSGy1OqVz5vZsrj2yiygHeX7wMxS036Mib/LuWrnkyk6VzWeBa4FBhCWB04pr7spdTXvsrRtq1j3/2em+zlXpbwF4lzVuAe4PK61ku5N4HgII6qA2RuYB2ohBbxMgKvZ/MzFuSpgZjOAm8q56zLgXkmfEKbNP2kDzzNH0juSPiMU55/PdazOZcvnwnLOOZcV78JyzjmXFU8gzjnnsuIJxDnnXFY8gTjnnMuKJxDnnHNZ8QTinHMuK55AnHPOZcUTiHPOuaz8P/nvh7kYiskwAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "color_coded = []\n",
    "for i in Return:\n",
    "    if i > 0:\n",
    "        color_coded.append('g')\n",
    "    else:\n",
    "        color_coded.append('r')\n",
    "\n",
    "fig, (ax1, ax2) = plt.subplots(2, sharex=True)\n",
    "fig.suptitle('Bitcoin Return Visualization')\n",
    "\n",
    "ax1.plot(my_data['date'], my_data['start_balance'])\n",
    "ax1.set(ylabel=\"Start Balance\")\n",
    "\n",
    "\n",
    "ax2.bar(my_data['date'], my_data['return'], color=color_coded)\n",
    "plt.axhline(y=0, color='black', linestyle='-')\n",
    "plt.xticks(rotation=45) \n",
    "plt.xlabel(\"Month\")\n",
    "plt.ylabel(\"Return\")\n",
    "pass"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4: Clinical disease data (40 pts)\n",
    "\n",
    "Your boss comes to you Monday morning and says “I figured out our next step; we are going to pivot from an online craft store and become a data center for genetic disease information! I found **ClinVar** which is a repository that contains expert curated data, and it is free for the taking. This is a gold mine! Look at the file and tell me what gene and mutation combinations are classified as dangerous.”\n",
    "\n",
    "Make sure that you only give your boss the dangerous mutations and include:\n",
    "\n",
    "1) Gene name\n",
    "\n",
    "2) Mutation ID number\n",
    "\n",
    "3) Mutation Position (chromosome & position)\n",
    "\n",
    "4) Mutation value (reference & alternate bases)\n",
    "\n",
    "5) Clinical significance (CLNSIG)\n",
    "\n",
    "6) Disease that is implicated\n",
    "\n",
    "**Requirements**\n",
    "\n",
    "1) The deliverables are the final result as a dataframe with a short discussion of any specifics. (that is, what data you would present to your boss with the explanation of your results)\n",
    "\n",
    "2) Limit your output to the first 100 harmful mutations and tell your boss how many total harmful mutations were found in the file\n",
    "\n",
    "3) Use the instructor-modified \"clinvar_final.txt\" at this link: https://drive.google.com/file/d/1Zps0YssoJbZHrn6iLte2RDLlgruhAX1s/view?usp=sharing This file was modified to be not exactly the same as 'standard' .vcf file to test your data parsing skills. **This is a large file so do NOT upload it into your github repo!**\n",
    "\n",
    "4) Replace missing values in the dataframe with: 'Not_Given'. Print or display this (including the Not_Given count) for the column `CLNSIG` by using pandas value_counts() function (https://pandas.pydata.org/docs/reference/api/pandas.Series.value_counts.html).\n",
    "\n",
    "5) State in your answer how you define harmful mutations\n",
    "\n",
    "**6) Do your best on getting to above requirements and submit whatever you do before the deadline. If your work is incomplete be sure to describe the blockers that got in your way and how you might get past them (if given more time).**\n",
    "\n",
    "7) You can use as many code blocks as you need. Please clean-up your code and make it readable for the graders!\n",
    "\n",
    "**Hints** \n",
    "* We do not expect you to have any medical knowledge to solve this problem; look at the data, read the documentation provided, and write down your assumptions!\n",
    "\n",
    "* Correct pseudocode will give you partial credit so start with that. \n",
    "\n",
    "* Map out which fields you want to extract: Are they in the same place every time? What strategy will you use to robustly extract and filter your data of interest? How do you plan to handle missing data?\n",
    "\n",
    "* A good way to start is to print out each line, then practice parsing them to see if you can recover the fields of interest\n",
    "\n",
    "* A starting solution for parsing .vcfs can be found here: https://gist.github.com/dceoy/99d976a2c01e7f0ba1c813778f9db744 This solution does **NOT** work due to the changes we've made but can be modified to work. As with any solution that needs modifications, it may take less time to make your own solution!\n",
    "\n",
    "* Filter out junk and lines with no mutation data. Just focus on the data your need to deliver to your boss. \n",
    "\n",
    "* Pandas and NumPy parsers correctly recognize the end of each line in in the ClinVar file.\n",
    "\n",
    "* The unit of observation of this dataset is one row per mutation.\n",
    "\n",
    "* This is similar to a task that one of us tackled at work. You can answer the question with the information provided below or using the (partial) data dictionary file at this link: https://drive.google.com/file/d/1lx9yHdlcqmU_OlHiTUXKC_LQDqYBypH_/view?usp=sharing. Our goal is to see that you can put together a sensible plan, implement a solid parsing strategy, and document and justify the decisions that you made."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### VCF file description (Summarized from version 4.1)\n",
    "\n",
    "```\n",
    "* The VCF specification:\n",
    "\n",
    "VCF is a text file format which contains meta-information lines, a header line, and then data lines each containing information about a position in the genome. The format also can contain genotype information on samples for each position.\n",
    "\n",
    "* Fixed fields:\n",
    "\n",
    "There are 8 fixed fields per record. All data lines are **tab-delimited**. In all cases, missing values are specified with a dot (‘.’). \n",
    "\n",
    "1. CHROM - chromosome number\n",
    "2. POS - position DNA nuceleotide count (bases) along the chromosome\n",
    "3. ID - The unique identifier for each mutation\n",
    "4. REF - reference base(s)\n",
    "5. ALT - alternate base(s)\n",
    "6. FILTER - filter status\n",
    "7. QUAL - quality\n",
    "8. INFO - a semicolon-separated series of keys with values in the format: <key>=<data>\n",
    "\n",
    "```\n",
    "### Applicable INFO field specifications\n",
    "\n",
    "```\n",
    "GENEINFO = <Gene name>\n",
    "CLNSIG =  <Clinical significance>\n",
    "CLNDN = <Disease name>\n",
    "```\n",
    "\n",
    "### Sample ClinVar data (vcf file format - not exactly the same as the file to download!)\n",
    "\n",
    "```\n",
    "##fileformat=VCFv4.1\n",
    "##fileDate=2019-03-19\n",
    "##source=ClinVar\n",
    "##reference=GRCh38\t\t\t\t\t\t\t\n",
    "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\n",
    "1\t949523\trs786201005\tC\tT\t.\t.\tGENEINFO=ISG15;CLNSIG=5\n",
    "1\t949696\trs672601345\tC\tCG\t.\t.\tGENEINFO=ISG15;CLNSIG=5;CLNDN=Cancer\n",
    "1\t949739\trs672601312\tG\tT\t.\t.\tGENEINFO=ISG15;CLNDBN=Cancer\n",
    "1\t955597\trs115173026\tG\tT\t.\t.\tGENEINFO=AGRN;CLNSIG=2; CLNDN=Cancer\n",
    "1\t955619\trs201073369\tG\tC\t.\t.\tGENEINFO=AGG;CLNDN=Heart_dis \n",
    "1\t957640\trs6657048\tC\tT\t.\t.\tGENEINFO=AGG;CLNSIG=3;CLNDN=Heart_dis \n",
    "1\t976059\trs544749044\tC\tT\t.\t.\tGENEINFO=AGG;CLNSIG=0;CLNDN=Heart_dis \n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>INFO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1014O42</td>\n",
       "      <td>475283</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_ESP=0.00546;AF_EXAC=0.00165;AF_TGP=0.00619;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1O14122</td>\n",
       "      <td>542074</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_ESP=0.00015;AF_EXAC=0.00010;ALLELEID=514926...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1014143</td>\n",
       "      <td>183381</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=181485;CLNDISDB=MedGen:C4015293,OMIM:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1014179</td>\n",
       "      <td>542075</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ALLELEID=514896;CLNDISDB=MedGen:C4015293,OMIM:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1014217</td>\n",
       "      <td>475278</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_ESP=0.00515;AF_EXAC=0.00831;AF_TGP=0.00339;...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM      POS      ID REF ALT FILTER QUAL  \\\n",
       "0      1  1014O42  475283   G   A      .    .   \n",
       "1      1  1O14122  542074   C   T      .    .   \n",
       "2      1  1014143  183381   C   T      .    .   \n",
       "3      1  1014179  542075   C   T      .    .   \n",
       "4      1  1014217  475278   C   T      .    .   \n",
       "\n",
       "                                                INFO  \n",
       "0  AF_ESP=0.00546;AF_EXAC=0.00165;AF_TGP=0.00619;...  \n",
       "1  AF_ESP=0.00015;AF_EXAC=0.00010;ALLELEID=514926...  \n",
       "2  ALLELEID=181485;CLNDISDB=MedGen:C4015293,OMIM:...  \n",
       "3  ALLELEID=514896;CLNDISDB=MedGen:C4015293,OMIM:...  \n",
       "4  AF_ESP=0.00515;AF_EXAC=0.00831;AF_TGP=0.00339;...  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#reading the file and converting it into DataFrames. \n",
    "\n",
    "import io\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "def read_vcf(path):\n",
    "    with open(path, 'r') as f:\n",
    "        lines = [l for l in f if not l.startswith('#')]\n",
    "    return pd.read_csv(\n",
    "        io.StringIO(''.join(lines)),\n",
    "        dtype={'#CHROM': str, 'POS': str, 'ID': str, 'REF': str, 'ALT': str,\n",
    "               'QUAL': str, 'FILTER': str, 'INFO': str},\n",
    "        sep='\\t'\n",
    "    ).rename(columns={'#CHROM': 'CHROM'})\n",
    "\n",
    "\n",
    "df = read_vcf(\"clinvar_final.txt\")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Getting CLNDN from INFO column. \n",
    "\n",
    "def parse_CLNDN(info: str):\n",
    "\n",
    "    split_data = info.split(\";\")\n",
    "    for i in split_data:\n",
    "        if \"CLNDN\" in i:\n",
    "            return i.split(\"=\")[-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Getting Gene info from INFO column. \n",
    "\n",
    "def parse_gene_info(info: str):\n",
    "    \n",
    "    split_data = info.split(\";\")\n",
    "    for i in split_data:\n",
    "        if \"GENEINFO\" in i:\n",
    "            return i.split(\"=\")[-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Getting CLNSIG from INFO column. \n",
    "\n",
    "def parse_CLNSIG(info: str):\n",
    "    \n",
    "    split_data = info.split(\";\")\n",
    "    for i in split_data:\n",
    "        if \"CLNSIG\" in i:\n",
    "            return i.split(\"=\")[-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>Mutation_ID_number</th>\n",
       "      <th>Mutation_Position</th>\n",
       "      <th>Mutation_value</th>\n",
       "      <th>Clinical_Significance</th>\n",
       "      <th>Disease_implicated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>475283</td>\n",
       "      <td>1014O42</td>\n",
       "      <td>GA</td>\n",
       "      <td>Benign</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>542074</td>\n",
       "      <td>1O14122</td>\n",
       "      <td>CT</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>183381</td>\n",
       "      <td>1014143</td>\n",
       "      <td>CT</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>542075</td>\n",
       "      <td>1014179</td>\n",
       "      <td>CT</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>475278</td>\n",
       "      <td>1014217</td>\n",
       "      <td>CT</td>\n",
       "      <td>Benign</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>402986</td>\n",
       "      <td>1014228</td>\n",
       "      <td>GA</td>\n",
       "      <td>Benign</td>\n",
       "      <td>not_specified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>571208</td>\n",
       "      <td>1014255</td>\n",
       "      <td>GA</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>568195</td>\n",
       "      <td>1014276</td>\n",
       "      <td>GA</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>161455</td>\n",
       "      <td>1014316</td>\n",
       "      <td>CCG</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>161454</td>\n",
       "      <td>1014359</td>\n",
       "      <td>GT</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene_name Mutation_ID_number Mutation_Position Mutation_value  \\\n",
       "0  ISG15:9636             475283           1014O42             GA   \n",
       "1  ISG15:9636             542074           1O14122             CT   \n",
       "2  ISG15:9636             183381           1014143             CT   \n",
       "3  ISG15:9636             542075           1014179             CT   \n",
       "4  ISG15:9636             475278           1014217             CT   \n",
       "5  ISG15:9636             402986           1014228             GA   \n",
       "6  ISG15:9636             571208           1014255             GA   \n",
       "7  ISG15:9636             568195           1014276             GA   \n",
       "8  ISG15:9636             161455           1014316            CCG   \n",
       "9  ISG15:9636             161454           1014359             GT   \n",
       "\n",
       "    Clinical_Significance                                 Disease_implicated  \n",
       "0                  Benign  Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "1  Uncertain_significance  Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "2              Pathogenic  Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "3  Uncertain_significance  Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "4                  Benign  Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "5                  Benign                                      not_specified  \n",
       "6  Uncertain_significance  Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "7  Uncertain_significance  Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "8              Pathogenic  Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "9              Pathogenic  Immunodeficiency_38_with_basal_ganglia_calcifi...  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# creating new DataFrame with Required columns. \n",
    "\n",
    "new_df = []\n",
    "\n",
    "for row in df.iterrows():\n",
    "    info = row[1][-1]\n",
    "    new_df.append(\n",
    "        {\n",
    "            \"Gene_name\": parse_gene_info(info),\n",
    "            \"Mutation_ID_number\": row[1][2],\n",
    "            \"Mutation_Position\": row[1][1],\n",
    "            \"Mutation_value\": row[1][3] + row[1][4],\n",
    "            \"Clinical_Significance\": parse_CLNSIG(info),\n",
    "            \"Disease_implicated\": parse_CLNDN(info)\n",
    "        }\n",
    "    )\n",
    "\n",
    "new_df = pd.DataFrame(new_df)\n",
    "new_df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False    100658\n",
       "True       1663\n",
       "Name: Clinical_Significance, dtype: int64"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# finding missing values\n",
    "\n",
    "new_df[\"Clinical_Significance\"].isnull().value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>Mutation_ID_number</th>\n",
       "      <th>Mutation_Position</th>\n",
       "      <th>Mutation_value</th>\n",
       "      <th>Clinical_Significance</th>\n",
       "      <th>Disease_implicated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>446</th>\n",
       "      <td>DVL1:1855</td>\n",
       "      <td>208189</td>\n",
       "      <td>1338183</td>\n",
       "      <td>TCTGGCGGGGGAGGGTAGGTGAGGGCCGCGGAGGGGCCTCCGGCGT...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>957</th>\n",
       "      <td>TNFRSF14:8764|TNFRSF14-AS1:115110</td>\n",
       "      <td>135349</td>\n",
       "      <td>2556714</td>\n",
       "      <td>AG</td>\n",
       "      <td>None</td>\n",
       "      <td>not_specified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>958</th>\n",
       "      <td>TNFRSF14:8764</td>\n",
       "      <td>133394</td>\n",
       "      <td>2559013</td>\n",
       "      <td>CT</td>\n",
       "      <td>None</td>\n",
       "      <td>not_specified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>959</th>\n",
       "      <td>TNFRSF14:8764</td>\n",
       "      <td>133395</td>\n",
       "      <td>2559074</td>\n",
       "      <td>CG</td>\n",
       "      <td>None</td>\n",
       "      <td>not_specified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>960</th>\n",
       "      <td>TNFRSF14:8764</td>\n",
       "      <td>133396</td>\n",
       "      <td>2559092</td>\n",
       "      <td>CT</td>\n",
       "      <td>None</td>\n",
       "      <td>not_specified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101815</th>\n",
       "      <td>PDCD10:11235</td>\n",
       "      <td>427927</td>\n",
       "      <td>167715673</td>\n",
       "      <td>TC</td>\n",
       "      <td>None</td>\n",
       "      <td>Cerebral_cavernous_malformations_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101816</th>\n",
       "      <td>PDCD10:11235</td>\n",
       "      <td>427909</td>\n",
       "      <td>167719197</td>\n",
       "      <td>TG</td>\n",
       "      <td>None</td>\n",
       "      <td>Cerebral_cavernous_malformations_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101818</th>\n",
       "      <td>PDCD10:11235</td>\n",
       "      <td>427920</td>\n",
       "      <td>167722405</td>\n",
       "      <td>AC</td>\n",
       "      <td>None</td>\n",
       "      <td>Cerebral_cavernous_malformations_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101819</th>\n",
       "      <td>PDCD10:11235</td>\n",
       "      <td>427926</td>\n",
       "      <td>167733204</td>\n",
       "      <td>GA</td>\n",
       "      <td>None</td>\n",
       "      <td>Cerebral_cavernous_malformations_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102314</th>\n",
       "      <td>PIK3CA:5290</td>\n",
       "      <td>376359</td>\n",
       "      <td>179210438</td>\n",
       "      <td>CT</td>\n",
       "      <td>None</td>\n",
       "      <td>Trabecular_adenocarcinoma</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1663 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                Gene_name Mutation_ID_number  \\\n",
       "446                             DVL1:1855             208189   \n",
       "957     TNFRSF14:8764|TNFRSF14-AS1:115110             135349   \n",
       "958                         TNFRSF14:8764             133394   \n",
       "959                         TNFRSF14:8764             133395   \n",
       "960                         TNFRSF14:8764             133396   \n",
       "...                                   ...                ...   \n",
       "101815                       PDCD10:11235             427927   \n",
       "101816                       PDCD10:11235             427909   \n",
       "101818                       PDCD10:11235             427920   \n",
       "101819                       PDCD10:11235             427926   \n",
       "102314                        PIK3CA:5290             376359   \n",
       "\n",
       "       Mutation_Position                                     Mutation_value  \\\n",
       "446              1338183  TCTGGCGGGGGAGGGTAGGTGAGGGCCGCGGAGGGGCCTCCGGCGT...   \n",
       "957              2556714                                                 AG   \n",
       "958              2559013                                                 CT   \n",
       "959              2559074                                                 CG   \n",
       "960              2559092                                                 CT   \n",
       "...                  ...                                                ...   \n",
       "101815         167715673                                                 TC   \n",
       "101816         167719197                                                 TG   \n",
       "101818         167722405                                                 AC   \n",
       "101819         167733204                                                 GA   \n",
       "102314         179210438                                                 CT   \n",
       "\n",
       "       Clinical_Significance                  Disease_implicated  \n",
       "446                     None                                None  \n",
       "957                     None                       not_specified  \n",
       "958                     None                       not_specified  \n",
       "959                     None                       not_specified  \n",
       "960                     None                       not_specified  \n",
       "...                      ...                                 ...  \n",
       "101815                  None  Cerebral_cavernous_malformations_3  \n",
       "101816                  None  Cerebral_cavernous_malformations_3  \n",
       "101818                  None  Cerebral_cavernous_malformations_3  \n",
       "101819                  None  Cerebral_cavernous_malformations_3  \n",
       "102314                  None           Trabecular_adenocarcinoma  \n",
       "\n",
       "[1663 rows x 6 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# there are 1663 missing values for Clinical Significance. \n",
    "# taking a look at missing values\n",
    "\n",
    "new_df[new_df[\"Clinical_Significance\"].isnull()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# replacing missing values with Not_Given as required. \n",
    "\n",
    "new_df.fillna(value=\"Not_Given\", axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False    102321\n",
       "Name: Clinical_Significance, dtype: int64"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# checking if there are still any missing values left after replacing them with Not_Given. \n",
    "# no missing values found now. \n",
    "\n",
    "new_df[\"Clinical_Significance\"].isnull().value_counts()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>Mutation_ID_number</th>\n",
       "      <th>Mutation_Position</th>\n",
       "      <th>Mutation_value</th>\n",
       "      <th>Clinical_Significance</th>\n",
       "      <th>Disease_implicated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>475283</td>\n",
       "      <td>1014O42</td>\n",
       "      <td>GA</td>\n",
       "      <td>Benign</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>542074</td>\n",
       "      <td>1O14122</td>\n",
       "      <td>CT</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>183381</td>\n",
       "      <td>1014143</td>\n",
       "      <td>CT</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>542075</td>\n",
       "      <td>1014179</td>\n",
       "      <td>CT</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>475278</td>\n",
       "      <td>1014217</td>\n",
       "      <td>CT</td>\n",
       "      <td>Benign</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>541170</td>\n",
       "      <td>1044019</td>\n",
       "      <td>GT</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>Myasthenic_syndrome,_congenital,_8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>501493</td>\n",
       "      <td>1044122</td>\n",
       "      <td>CT</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>Not_Given</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>128293</td>\n",
       "      <td>1044134</td>\n",
       "      <td>CG</td>\n",
       "      <td>Benign</td>\n",
       "      <td>Myasthenic_syndrome,_congenital,_8|not_specified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>541180</td>\n",
       "      <td>1044137</td>\n",
       "      <td>TA</td>\n",
       "      <td>Likely_benign</td>\n",
       "      <td>Myasthenic_syndrome,_congenital,_8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>578507</td>\n",
       "      <td>1044147</td>\n",
       "      <td>GA</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>Myasthenic_syndrome,_congenital,_8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Gene_name Mutation_ID_number Mutation_Position Mutation_value  \\\n",
       "0    ISG15:9636             475283           1014O42             GA   \n",
       "1    ISG15:9636             542074           1O14122             CT   \n",
       "2    ISG15:9636             183381           1014143             CT   \n",
       "3    ISG15:9636             542075           1014179             CT   \n",
       "4    ISG15:9636             475278           1014217             CT   \n",
       "..          ...                ...               ...            ...   \n",
       "95  AGRN:375790             541170           1044019             GT   \n",
       "96  AGRN:375790             501493           1044122             CT   \n",
       "97  AGRN:375790             128293           1044134             CG   \n",
       "98  AGRN:375790             541180           1044137             TA   \n",
       "99  AGRN:375790             578507           1044147             GA   \n",
       "\n",
       "     Clinical_Significance                                 Disease_implicated  \n",
       "0                   Benign  Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "1   Uncertain_significance  Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "2               Pathogenic  Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "3   Uncertain_significance  Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "4                   Benign  Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "..                     ...                                                ...  \n",
       "95  Uncertain_significance                 Myasthenic_syndrome,_congenital,_8  \n",
       "96  Uncertain_significance                                          Not_Given  \n",
       "97                  Benign   Myasthenic_syndrome,_congenital,_8|not_specified  \n",
       "98           Likely_benign                 Myasthenic_syndrome,_congenital,_8  \n",
       "99  Uncertain_significance                 Myasthenic_syndrome,_congenital,_8  \n",
       "\n",
       "[100 rows x 6 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Taking a look at our DataFrame\n",
    "new_df.head(100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>Mutation_ID_number</th>\n",
       "      <th>Mutation_Position</th>\n",
       "      <th>Mutation_value</th>\n",
       "      <th>Clinical_Significance</th>\n",
       "      <th>Disease_implicated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>183381</td>\n",
       "      <td>1014143</td>\n",
       "      <td>CT</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>161455</td>\n",
       "      <td>1014316</td>\n",
       "      <td>CCG</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>161454</td>\n",
       "      <td>1014359</td>\n",
       "      <td>GT</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>243036</td>\n",
       "      <td>1022225</td>\n",
       "      <td>GA</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Congenital_myasthenic_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>243037</td>\n",
       "      <td>1022313</td>\n",
       "      <td>AT</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Congenital_myasthenic_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1789</th>\n",
       "      <td>ESPN:83715</td>\n",
       "      <td>4420</td>\n",
       "      <td>6451926</td>\n",
       "      <td>AC</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Deafness,_without_vestibular_involvement,_auto...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1794</th>\n",
       "      <td>ESPN:83715</td>\n",
       "      <td>4422</td>\n",
       "      <td>6452092</td>\n",
       "      <td>GA</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Deafness,_without_vestibular_involvement,_auto...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1795</th>\n",
       "      <td>ESPN:83715</td>\n",
       "      <td>560535</td>\n",
       "      <td>6457223</td>\n",
       "      <td>TGGAGGCGGGACCTCCTGCT</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Usher_syndrome,_type_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1798</th>\n",
       "      <td>ESPN:83715</td>\n",
       "      <td>562076</td>\n",
       "      <td>6460076</td>\n",
       "      <td>ACA</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Deafness,_autosomal_recessive_36,_with_or_with...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1867</th>\n",
       "      <td>PLEKHG5:57449</td>\n",
       "      <td>566096</td>\n",
       "      <td>6468294</td>\n",
       "      <td>GA</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Distal_spinal_muscular_atrophy,_autosomal_rece...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Gene_name Mutation_ID_number Mutation_Position  \\\n",
       "2        ISG15:9636             183381           1014143   \n",
       "8        ISG15:9636             161455           1014316   \n",
       "9        ISG15:9636             161454           1014359   \n",
       "24      AGRN:375790             243036           1022225   \n",
       "26      AGRN:375790             243037           1022313   \n",
       "...             ...                ...               ...   \n",
       "1789     ESPN:83715               4420           6451926   \n",
       "1794     ESPN:83715               4422           6452092   \n",
       "1795     ESPN:83715             560535           6457223   \n",
       "1798     ESPN:83715             562076           6460076   \n",
       "1867  PLEKHG5:57449             566096           6468294   \n",
       "\n",
       "            Mutation_value Clinical_Significance  \\\n",
       "2                       CT            Pathogenic   \n",
       "8                      CCG            Pathogenic   \n",
       "9                       GT            Pathogenic   \n",
       "24                      GA            Pathogenic   \n",
       "26                      AT            Pathogenic   \n",
       "...                    ...                   ...   \n",
       "1789                    AC            Pathogenic   \n",
       "1794                    GA            Pathogenic   \n",
       "1795  TGGAGGCGGGACCTCCTGCT            Pathogenic   \n",
       "1798                   ACA            Pathogenic   \n",
       "1867                    GA            Pathogenic   \n",
       "\n",
       "                                     Disease_implicated  \n",
       "2     Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "8     Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "9     Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "24                       Congenital_myasthenic_syndrome  \n",
       "26                       Congenital_myasthenic_syndrome  \n",
       "...                                                 ...  \n",
       "1789  Deafness,_without_vestibular_involvement,_auto...  \n",
       "1794  Deafness,_without_vestibular_involvement,_auto...  \n",
       "1795                             Usher_syndrome,_type_1  \n",
       "1798  Deafness,_autosomal_recessive_36,_with_or_with...  \n",
       "1867  Distal_spinal_muscular_atrophy,_autosomal_rece...  \n",
       "\n",
       "[100 rows x 6 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# I am going to only pick the diseases that are considered pathogenic under Clinical Significance.\n",
    "# Printing first 100 rows.\n",
    "\n",
    "df1 = new_df[new_df['Clinical_Significance'] == 'Pathogenic']\n",
    "df1.head(100)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>Mutation_ID_number</th>\n",
       "      <th>Mutation_Position</th>\n",
       "      <th>Mutation_value</th>\n",
       "      <th>Disease_implicated</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Clinical_Significance</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Pathogenic</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>183381</td>\n",
       "      <td>1014143</td>\n",
       "      <td>CT</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pathogenic</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>161455</td>\n",
       "      <td>1014316</td>\n",
       "      <td>CCG</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pathogenic</th>\n",
       "      <td>ISG15:9636</td>\n",
       "      <td>161454</td>\n",
       "      <td>1014359</td>\n",
       "      <td>GT</td>\n",
       "      <td>Immunodeficiency_38_with_basal_ganglia_calcifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pathogenic</th>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>243036</td>\n",
       "      <td>1022225</td>\n",
       "      <td>GA</td>\n",
       "      <td>Congenital_myasthenic_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pathogenic</th>\n",
       "      <td>AGRN:375790</td>\n",
       "      <td>243037</td>\n",
       "      <td>1022313</td>\n",
       "      <td>AT</td>\n",
       "      <td>Congenital_myasthenic_syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pathogenic</th>\n",
       "      <td>ESPN:83715</td>\n",
       "      <td>4420</td>\n",
       "      <td>6451926</td>\n",
       "      <td>AC</td>\n",
       "      <td>Deafness,_without_vestibular_involvement,_auto...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pathogenic</th>\n",
       "      <td>ESPN:83715</td>\n",
       "      <td>4422</td>\n",
       "      <td>6452092</td>\n",
       "      <td>GA</td>\n",
       "      <td>Deafness,_without_vestibular_involvement,_auto...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pathogenic</th>\n",
       "      <td>ESPN:83715</td>\n",
       "      <td>560535</td>\n",
       "      <td>6457223</td>\n",
       "      <td>TGGAGGCGGGACCTCCTGCT</td>\n",
       "      <td>Usher_syndrome,_type_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pathogenic</th>\n",
       "      <td>ESPN:83715</td>\n",
       "      <td>562076</td>\n",
       "      <td>6460076</td>\n",
       "      <td>ACA</td>\n",
       "      <td>Deafness,_autosomal_recessive_36,_with_or_with...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pathogenic</th>\n",
       "      <td>PLEKHG5:57449</td>\n",
       "      <td>566096</td>\n",
       "      <td>6468294</td>\n",
       "      <td>GA</td>\n",
       "      <td>Distal_spinal_muscular_atrophy,_autosomal_rece...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           Gene_name Mutation_ID_number Mutation_Position  \\\n",
       "Clinical_Significance                                                       \n",
       "Pathogenic                ISG15:9636             183381           1014143   \n",
       "Pathogenic                ISG15:9636             161455           1014316   \n",
       "Pathogenic                ISG15:9636             161454           1014359   \n",
       "Pathogenic               AGRN:375790             243036           1022225   \n",
       "Pathogenic               AGRN:375790             243037           1022313   \n",
       "...                              ...                ...               ...   \n",
       "Pathogenic                ESPN:83715               4420           6451926   \n",
       "Pathogenic                ESPN:83715               4422           6452092   \n",
       "Pathogenic                ESPN:83715             560535           6457223   \n",
       "Pathogenic                ESPN:83715             562076           6460076   \n",
       "Pathogenic             PLEKHG5:57449             566096           6468294   \n",
       "\n",
       "                             Mutation_value  \\\n",
       "Clinical_Significance                         \n",
       "Pathogenic                               CT   \n",
       "Pathogenic                              CCG   \n",
       "Pathogenic                               GT   \n",
       "Pathogenic                               GA   \n",
       "Pathogenic                               AT   \n",
       "...                                     ...   \n",
       "Pathogenic                               AC   \n",
       "Pathogenic                               GA   \n",
       "Pathogenic             TGGAGGCGGGACCTCCTGCT   \n",
       "Pathogenic                              ACA   \n",
       "Pathogenic                               GA   \n",
       "\n",
       "                                                      Disease_implicated  \n",
       "Clinical_Significance                                                     \n",
       "Pathogenic             Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "Pathogenic             Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "Pathogenic             Immunodeficiency_38_with_basal_ganglia_calcifi...  \n",
       "Pathogenic                                Congenital_myasthenic_syndrome  \n",
       "Pathogenic                                Congenital_myasthenic_syndrome  \n",
       "...                                                                  ...  \n",
       "Pathogenic             Deafness,_without_vestibular_involvement,_auto...  \n",
       "Pathogenic             Deafness,_without_vestibular_involvement,_auto...  \n",
       "Pathogenic                                        Usher_syndrome,_type_1  \n",
       "Pathogenic             Deafness,_autosomal_recessive_36,_with_or_with...  \n",
       "Pathogenic             Distal_spinal_muscular_atrophy,_autosomal_rece...  \n",
       "\n",
       "[100 rows x 5 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Setting Clinical Significance as index\n",
    "# Printing 100 rows. \n",
    "\n",
    "df2 = df1.set_index('Clinical_Significance')\n",
    "df2.head(100)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After parsing and cleaning my data, my only assumption was that only the diseases that are considered pathogenic under clinical significance are caused by dangerous mutations. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Taking a preliminary look at ClinVar database, I was able to clean up the database and turn it into readable information. I created a table in which all mutations that can cause diseases and are considered Pathogenic under Clinical Significance are listed.\n",
    "\n",
    "As you can see above, I only printed the first 100 pathogenic mutations with specific gene numbers, mutation ID numbers, mutation positions, mutation values, and the specific diseases they may cause. \n",
    "\n",
    "If our company decides to become a data center for genetic disease information, I would be more than happy to assemble a team in which we will dive deep into this data and extract useful information. "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
